File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& m) {
        int l = 0, r = m[0].size(), t = 0, b = m.size();
        vector<int> res;
        
        while(l < r && t < b) {
            for(int i = l; i < r; i++) {
                res.push_back(m[t][i]);
            }
            t++;
            for(int i = t; i < b; i++) {
                res.push_back(m[i][r - 1]);
            }
            r--;
            if(!(l < r && t < b)) break;
            
            for(int i = r - 1; i >= l; i--) {
                res.push_back(m[b - 1][i]);
            }
            b--;
            for(int i = b - 1; i >= t; i--) {
                res.push_back(m[i][l]);
            }
            l++;
        }
        
        return res;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        l, r = 0, len(matrix[0])
        t, b = 0,len(matrix)
        
        res = []

        while l < r and t < b:
            # left to right
            for i in range(l, r):
                res.append(matrix[t][i])
            # done with top row
            t += 1

            # go through right col
            for i in range(t, b):
                res.append(matrix[i][r-1])

            # done with rightmost col
            r -= 1

            if not(l < r and t < b):
                break

            # go through the bottom col
            for i in range(r-1, l-1, -1):
                res.append(matrix[b-1][i])
            
            # done with the bottom row
            b -= 1

            # go through the left col
            for i in range(b-1, t-1, -1):
                res.append(matrix[i][l])
            
            # done with the lefft col
            l += 1
        return res